### PR TITLE
Add support for Tag triggers in GOGS/Gitea Webhook service

### PR DESCRIPTION
**Pull request by crrobinson14**
This PR addresses https://github.com/bitrise-io/bitrise-webhooks/issues/30 with two changes:

1. Adds support for Tag triggers, which are sent as `ref_type: "tag", ref: "TAGNAME"`.
1. Adds doc references to Gitea, which the GOGS handler is compatible with (since Gitea is a GOGS fork).

### CODE REVIEW

File: service/hook/gogs/gogs.go [hunk 1]
```diff
@@ -82,17 +82,39 @@ func transformCodePushEvent(codePushEvent CodePushEventModel) hookCommon.Transfo
 
 	if !isLastCommitFound {
 		return hookCommon.TransformResultModel{
-			Error: errors.New("The commit specified by 'after' was not included in the 'commits' array - no match found"),
+			Error:      errors.New("The commit specified by 'after' was not included in the 'commits' array - no match found"),
+			ShouldSkip: true,
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: why did you change this to `ShouldSkip: true`?

This will answer with a "200" response, not with an error one.

File: service/hook/gogs/gogs.go [hunk 2]
```diff
@@ -82,17 +82,39 @@ func transformCodePushEvent(codePushEvent CodePushEventModel) hookCommon.Transfo
...
+		}
+	}
+
+	if len(lastCommit.CommitHash) == 0 {
+		return hookCommon.TransformResultModel{
+			Error:      fmt.Errorf("Missing commit hash"),
+			ShouldSkip: true,
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: should this return a "200 - ignoring" or an error instead?

File: service/hook/gogs/gogs_test.go [hunk 3]
```diff
@@ -190,7 +177,28 @@ func Test_HookProvider_TransformRequest(t *testing.T) {
 		}, hookTransformResult.TriggerAPIParams)
 	}
 
-	t.Log("Unsuported Content-Type")
+	t.Log("Tag Push - should be handled")
```
Comment (viktorbenei) [inline service/hook/gogs/gogs_test.go]: LGTM, thanks for the test! :+1:  :)

File: service/hook/gogs/gogs_test.go [hunk 4]
```diff
@@ -122,28 +122,9 @@ func Test_transformCodePushEvent(t *testing.T) {
...
 			},
 		}
-		hookTransformResult := transformCodePushEvent(codePush)
+		hookTransformResult := transformPushEvent(codePush)
 		require.EqualError(t, hookTransformResult.Error, "The commit specified by 'after' was not included in the 'commits' array - no match found")
-		require.False(t, hookTransformResult.ShouldSkip)
-		require.Nil(t, hookTransformResult.TriggerAPIParams)
-	}
-
-	t.Log("Not a head ref")
```
Comment (viktorbenei) [inline service/hook/gogs/gogs_test.go]: Why did you remove this test?

File: service/hook/gogs/gogs_test.go [hunk 5]
```diff
@@ -122,28 +122,9 @@ func Test_transformCodePushEvent(t *testing.T) {
 				},
 			},
 		}
-		hookTransformResult := transformCodePushEvent(codePush)
+		hookTransformResult := transformPushEvent(codePush)
 		require.EqualError(t, hookTransformResult.Error, "The commit specified by 'after' was not included in the 'commits' array - no match found")
-		require.False(t, hookTransformResult.ShouldSkip)
```
Comment (viktorbenei) [inline service/hook/gogs/gogs_test.go]: I'd leave these tests here, it helps with code revisions, to ensure compatibility, or to show when these do change

File: service/hook/gogs/gogs.go [hunk 6]
```diff
@@ -104,35 +126,44 @@ func (hp HookProvider) TransformRequest(r *http.Request) hookCommon.TransformRes
...
-		if err := json.NewDecoder(r.Body).Decode(&codePushEvent); err != nil {
+	if eventID == pushEventID {
+		var pushEvent PushEventModel
+		if err := json.NewDecoder(r.Body).Decode(&pushEvent); err != nil {
+			return hookCommon.TransformResultModel{Error: fmt.Errorf("Failed to parse request body: %s", err)}
+		}
+
+		return transformPushEvent(pushEvent)
+
+	} else if eventID == tagEventID {
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: I think this is not enough, I'm almost sure that when you create a new branch that's a `create` event too.

File: service/hook/gogs/gogs.go (hunk 1 above)
Comment (crrobinson14) [inline service/hook/gogs/gogs.go]: yes

File: service/hook/gogs/gogs.go (hunk 2 above)
Comment (crrobinson14) [inline service/hook/gogs/gogs.go]: No, because it basically means it's an invalid message, maybe from something that "looks like, but isn't, Gogs". Gogs should always send this hash... but if it doesn't, obviously we need to treat that as an error. I removed the SholdSkip tho.

File: service/hook/gogs/gogs.go (hunk 6 above)
Comment (crrobinson14) [inline service/hook/gogs/gogs.go]: Fixed in this next commit.

File: service/hook/gogs/gogs_test.go [hunk 7]
```diff
@@ -122,30 +122,11 @@ func Test_transformCodePushEvent(t *testing.T) {
...
 			},
 		}
-		hookTransformResult := transformCodePushEvent(codePush)
+		hookTransformResult := transformPushEvent(codePush)
 		require.EqualError(t, hookTransformResult.Error, "The commit specified by 'after' was not included in the 'commits' array - no match found")
 		require.False(t, hookTransformResult.ShouldSkip)
 		require.Nil(t, hookTransformResult.TriggerAPIParams)
 	}
-
-	t.Log("Not a head ref")
```
Comment (viktorbenei) [inline service/hook/gogs/gogs_test.go]: Why did you remove this test?

File: README.md [hunk 8]
```diff
@@ -92,15 +92,17 @@ a build will be triggered (if you have Trigger mapping defined for the event(s)
...
 1. Select `Webhooks`, `Add Webhook`, then `Gogs`.
 1. Specify the `bitrise-webhooks` URL (`.../h/gogs/BITRISE-APP-SLUG/BITRISE-APP-API-TOKEN`) in the `Payload URL` field.
 1. Set the `Content Type` to `application/json`.
 1. A Secret is not required at this time.
-1. Set the trigger to be fired on `Just the push event`
+1. Set the trigger to be fired on either `Just the push event` or `Let me choose what I need` and select
+`Create` and `Push`. (Pull request triggers are not supported at this time.)
 1. Save the Webhook.
 
 That's all! The next time you __push code__
```
Comment (viktorbenei) [inline README.md]: I'd add ` or __push a new tag__`, just like eg. https://github.com/bitrise-io/bitrise-webhooks/pull/43/files#diff-04c6e90faac2675aa89e2176d2eec7d8R89

File: service/hook/gogs/gogs.go [hunk 9]
```diff
@@ -33,14 +34,21 @@ type CommitModel struct {
...
+type PushEventModel struct {
 	Secret      string        `json:"secret"`
 	Ref         string        `json:"ref"`
 	CheckoutSHA string        `json:"after"`
 	Commits     []CommitModel `json:"commits"`
 }
 
+// CreateEventModel ...
+type CreateEventModel struct {
+	Secret  string `json:"secret"`
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: As secret is not handled anywhere, this could also be removed

File: service/hook/gogs/gogs.go [hunk 10]
```diff
@@ -33,14 +34,21 @@ type CommitModel struct {
 	CommitMessage string `json:"message"`
 }
 
-// CodePushEventModel ...
-type CodePushEventModel struct {
+// PushEventModel ...
+type PushEventModel struct {
 	Secret      string        `json:"secret"`
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: As secret is not handled anywhere, this could also be removed

File: service/hook/gogs/gogs.go [hunk 11]
```diff
@@ -61,19 +69,11 @@ func detectContentTypeAndEventID(header http.Header) (string, string, error) {
 	return contentType, eventID, nil
 }
 
-func transformCodePushEvent(codePushEvent CodePushEventModel) hookCommon.TransformResultModel {
-	if !strings.HasPrefix(codePushEvent.Ref, "refs/heads/") {
```
Comment (viktorbenei) [inline service/hook/gogs/gogs.go]: This could be kept, related to https://github.com/bitrise-io/bitrise-webhooks/pull/43/files#r101161888

**Comment by viktorbenei** [comment]
@crrobinson14 just found this PR again. Checking it right now, my only complaint is that you did remove a test case: https://github.com/bitrise-io/bitrise-webhooks/pull/43/files#diff-5261bf95fc831f9ef14308209e2340c2L131

Was there a specific use case / issue why you removed that test?

**Comment by crrobinson14** [comment]
Sorry, @viktorbenei, I have only certain time windows for working on projects like this and I just hit my time limit with this one. I'll see if I can find a slot this week to finish it off. I have to refresh my memory on the change.

**Comment by viktorbenei** [comment]
Sure, no problem at all. If you won't have the time I can also send a PR, based on yours, just let me know

**Comment by viktorbenei** [comment]
Closing as this is quite outdated now + has a merge conflict.

@crrobinson14 if you have some time feel free to open a new PR and let us know if we can help with anything!
